### DISCUSSION TITLE
What Prompt Toolkit-based keybindings are you all using?

### DISCUSSION THREAD

**Discussion by yggdr** [xonsh/xonsh#5671] [Q&A]
I was wondering what keybindings you are all using. I'm trying to be more conscious about my workflows on the command line, and how the possibilities that prompt toolkit gives us can make them more fluent. There are already are some nice defaults compared to other shells (like no strange stuff being inserted on pressing C-Arrow, or DEL), but there is still room for improvement.

My example is an expansion on the idea of  plugin for other shells that allows you to add "sudo " to the front of the current line:

```xonsh
from prompt_toolkit.document import Document as _Document
from prompt_toolkit.key_binding.key_processor import KeyPress as _KeyPress
from prompt_toolkit.keys import Keys as _Keys
from xonsh.events import events as _events

@_events.on_ptk_create
def _custom_keybindings(bindings, **kw):

    ...  # other, smaller stuff

    def del_first_words(n, event, space_at_start=False):
        orig_text = event.current_buffer.text
        *_, new_text = event.current_buffer.text.split(maxsplit=int(n))
        if space_at_start:
            new_text = f" {new_text}"
        if new_text != orig_text:
            pos = event.current_buffer.cursor_position - len(orig_text) + len(new_text)
            # Setting new document because using cursor_left sometimes is off by n
            event.current_buffer.document = _Document(new_text, pos)

    mods = {
        'f': ('f"{', '}"'),
        'u': ('sudo ', ''),
        'w': ('pkg which $(', ')'),
        'W': ('pkg which $(which ', ')'),
    }

    @bindings.add(_Keys.ControlO, _Keys.Any)
    def modify_input(event):
        match event.key_sequence[1]:
            case _KeyPress(key=n) if n.isdigit():
                del_first_words(n, event)
            case _KeyPress(key='o'):
                del_first_words(1, event)
            case _KeyPress(key=_Keys.ControlO):
                del_first_words(1, event, space_at_start=True)
                event.current_buffer.cursor_position = 0
            case _KeyPress(key=n) if n in mods:
                pre, post = mods[n]
                event.current_buffer.text = f'{pre}{event.current_buffer.text}{post}'
                event.current_buffer.cursor_right(len(pre))
```

`C-O`, followed by a number deletes that many words from the start of the line, `C-O O` deletes one word.
`C-O C-O` deletes one word at the start, inserts a space at the beginning, then moves the cursor there, then you can correct a typo in the program you wanted to call, or invoke e.g. a different editor with the same set of files.
`C-O f` changes `stuff` to `f"{stuff}"` as a helper for printing out python variables inside xonsh
`C-O u` inserts `sudo ` at the start.
When I experiment with small helper programs I tend to want to try them out in my main environment instead of a jail or VM, so often I wonder how/where an executable comes from. Hence why I use `which some-executable` often. My package manager (FreeBSD's pkg) provides its own `pkg which some-path` subcommand that tells me which package installed a certain file. That's what the last two keybindings `C-O w` and `C-O W` help with.

I'm looking forward to hearing some more ideas you have come up with :)

**Comment by anki-code** [comment]
It will be cool to have a [xontrib](https://github.com/topics/xontrib) with all of this e.g. based on https://github.com/xonsh/xontrib-template.

**Reply by yggdr** [reply]
Interesting idea. If there are a few more ideas for key bindings incoming, I could see this becoming a kind of small framework for certain types of keybindings.
